src/Spd.Resource.Repository/BizLicApplication/BizLicApplicationRepository.cs

```diff
@@ -162,43 +166,62 @@ private void LinkOrganization(Guid? accountId, spd_application app)
         }
     }
 
-    private void LinkPrivateInvestigator(SwlContactInfo privateInvestigatorInfo, spd_application app)
+    private spd_businesscontact UpsertPrivateInvestigator(PrivateInvestigatorSwlContactInfo privateInvestigatorInfo, spd_application app)
     {
-        if (privateInvestigatorInfo.LicenceId == null)
-            return;
+        spd_businesscontact? bizContact = null;
+        DeletePrivateInvestigatorLink(app);
+        Guid? bizContactId = privateInvestigatorInfo?.BizContactId;
 
-        // Related licence already linked
-        if (app.spd_application_spd_licence_manager.Any(l => l.spd_licenceid == privateInvestigatorInfo.LicenceId))
-            return;
+        if (bizContactId == null)
+        {
+            bizContact = _mapper.Map<spd_businesscontact>(privateInvestigatorInfo);
+            bizContact.spd_businesscontactid = Guid.NewGuid();
+            _context.AddTospd_businesscontacts(bizContact);
+            AddPrivateInvestigatorLink(bizContact, app);
+        }
+        else
+        {
+            bizContact = _context.spd_businesscontacts.Where(b => b.spd_businesscontactid == bizContactId).FirstOrDefault();
 
-        spd_licence? licence = _context.spd_licences
-            .Where(l => l.spd_licenceid == privateInvestigatorInfo.LicenceId)
-            .Where(a => a.statecode == DynamicsConstants.StateCode_Active)
-            .FirstOrDefault();
+            if (bizContact == null)
+                throw new ArgumentException("business contact not found");
 
-        if (licence == null)
-            throw new ArgumentException("investigator licence info not found");
+            _mapper.Map(privateInvestigatorInfo, bizContact);
+            _context.UpdateObject(bizContact);
 
-        _context.AddLink(app, nameof(spd_application.spd_application_spd_licence_manager), licence);
+            // Must save changes before adding link that has navigation relationship, otherwise the transaction fails
+            _context.SaveChanges();
+            AddPrivateInvestigatorLink(bizContact, app);
+        }
+        return bizContact;
     }
 
-    private void DeletePrivateInvestigatorLink(SwlContactInfo privateInvestigatorInfo, spd_application app)
+    private void AddPrivateInvestigatorLink(spd_businesscontact bizContact, spd_application app)
     {
-        if (privateInvestigatorInfo.LicenceId == null)
-            return;
+        _context.AddLink(bizContact, nameof(spd_application.spd_businesscontact_spd_application), app);
+
+        var position = _context.LookupPosition(PositionEnum.PrivateInvestigatorManager.ToString());
+
+        if (position != null)
+            _context.AddLink(position, nameof(spd_businesscontact.spd_position_spd_businesscontact), bizContact);
+    }
+
+    private void DeletePrivateInvestigatorLink(spd_application app)
+    {
+        spd_businesscontact? bizContact = app.spd_businesscontact_spd_application.FirstOrDefault();
```

Comment:
one application can have many bizContact, how do you know you delete the priviateInvestigator one?

Reply:
Nice catch, I adjusted the query to make sure the record that has the private investigator is pulled